moveit_ros/moveit_servo/src/utils/common.cpp

```diff
@@ -138,53 +146,112 @@ geometry_msgs::msg::Pose poseFromCartesianDelta(const Eigen::VectorXd& delta_x,
+    point.positions.reserve(num_joints);
+    point.velocities.reserve(num_joints);
+    point.accelerations.reserve(num_joints);
+    if (servo_params.publish_joint_positions)
+    {
+      for (const auto& pos : state.positions)
+      {
+        point.positions.emplace_back(pos);
+      }
+    }
+    if (servo_params.publish_joint_velocities)
+    {
+      for (const auto& vel : state.velocities)
+      {
+        point.velocities.emplace_back(vel);
+      }
+    }
+    if (servo_params.publish_joint_accelerations)
+    {
+      for (const auto& acc : state.accelerations)
+      {
+        point.accelerations.emplace_back(acc);
+      }
+    }
+    point.time_from_start = state.time_stamp - joint_trajectory.header.stamp;
+    joint_trajectory.points.emplace_back(point);
+  };
+
+  for (size_t i = 0; i < joint_cmd_rolling_window.size() - 1; ++i)
   {
-    point.positions.resize(num_joints);
-    point.velocities.resize(num_joints);
-    point.accelerations.resize(num_joints);
+    add_point(joint_trajectory, joint_cmd_rolling_window[i]);
   }
 
-  // Set the fields of trajectory point based on which fields are requested.
-  // Some controllers check that acceleration data is non-empty, even if accelerations are not used
-  // Send all zeros (joint_state.accelerations is a vector of all zeros).
+  return joint_trajectory;
+}
 
-  if (servo_params.publish_joint_positions)
+void updateSlidingWindow(KinematicState& next_joint_state, std::deque<KinematicState>& joint_cmd_rolling_window,
+                         double max_expected_latency, const rclcpp::Time& cur_time)
+{
+  // remove commands older than current time minus the length of the sliding window
+  next_joint_state.time_stamp = cur_time + rclcpp::Duration::from_seconds(max_expected_latency);
+  const auto active_time_window = rclcpp::Duration::from_seconds(max_expected_latency);
+  while (!joint_cmd_rolling_window.empty() &&
+         joint_cmd_rolling_window.front().time_stamp < (cur_time - active_time_window))
   {
-    for (size_t i = 0; i < num_joints; ++i)
-    {
-      point.positions[i] = joint_state.positions[i];
-    }
+    joint_cmd_rolling_window.pop_front();
   }
-  if (servo_params.publish_joint_velocities)
+
+  // remove commands at end of window if timestamp is the same as current command
+  while (!joint_cmd_rolling_window.empty() && next_joint_state.time_stamp == joint_cmd_rolling_window.back().time_stamp)
   {
-    for (size_t i = 0; i < num_joints; ++i)
-    {
-      point.velocities[i] = joint_state.velocities[i];
-    }
+    joint_cmd_rolling_window.pop_back();
   }
-  if (servo_params.publish_joint_accelerations)
+
+  // update velocity: the velocity has the potential to dramatically influence interpolation of splines and causes large
+  // overshooting. To alleviate this effect, the target velocity will be set to zero if the motion changes direction,
+  // otherwise, it will calculate the forward and backward finite difference velocities and choose the minimum.
```

Comment:
No, there isn't a reference algorithm for this. I would prefer to have a spline method like PCHIP that doesn't have so much overshoot like regular cubic splines. Setting the velocity boundary conditions does help with overshooting. I think this could use some work. 

I think we can do better here. One problem is that Euler integration is being used to calculate the position and velocity of the next state given the current state and command. Since the integration isn't completely accurate, it can affect the boundary conditions. However, if we use a constant acceleration, then we can use the analytical integral solution. If the analytical solution is used, then we probably won't need the special checks in this method.